### ISSUE TITLE
ZIO Test: Add effectful suites

### ISSUE THREAD

**Issue by jdegoes** [zio/zio#1693]
This ticket is to track introduction of effectful suites into ZIO Test.

The end goal should be:

 - The new ability to provide an entire suite its dependency, a single time (while still retaining the ability to do this per test, as now)
 - Possibly, the ability to construct a suite dynamically from runtime data (I think this is unavoidable given (1), and if so, we should just embrace it)

Anything else? /cc  @adamgfraser  @iravid

**Comment by adamgfraser** [comment]
@jdegoes Completely agree. The latest version of #1664 goes quite very in that direction so I think if we refine that we should be able to more forward on this quite quickly..

**Comment by jdegoes** [comment]
One way to do this:

```scala
  final case class SuiteCase[+L, +A](label: L, specs: Vector[A], exec: Option[ExecutionStrategy])
      extends SpecCase[L, Nothing, A]
  final case class TestCase[+L, +T](label: L, test: T) extends SpecCase[L, T, Nothing]
```

Add another case:

```scala
  final case class SuiteCase[+L, +A](label: L, specs: Vector[A], exec: Option[ExecutionStrategy])
      extends SpecCase[L, Nothing, A]
  final case class TestCase[+L, +T](label: L, test: T) extends SpecCase[L, T, Nothing]
  final case class EffectCase[+L, +T, +A](effect: UIO[SpecCase[L, T, A]]) extends SpecCase[L, T, A]
```

Note that to be maximally useful (including solving (1) above), the effect could not be `UIO`, it would have to be `ZIO`, which means propagating `R` and `E` into `SpecCase`.

**Comment by jdegoes** [comment]
@adamgfraser Ah, I had no idea you were still working on that. Will take a look!

**Comment by jdegoes** [comment]
@adamgfraser Btw, with the above structure, note how it's possible to "pull" the effect out of `T` and place it directly in `SpecCase`.  That could make `T` just a test result.

**Comment by adamgfraser** [comment]
@jdegoes Yes that is very nice. I think it might also be possible to describe "pure" specs if we make `SuiteCase` and `TestCase` a subset of `PureCase`. Then we could describe specs that didn't involve any effects, which in particular could be nice for `ExecutedSpec`.

**Comment by adamgfraser** [comment]
Closed by #1664.
